### PR TITLE
add trace_id in logs to identify all logs related to a request/error

### LINKED ISSUE #1981: Add id in logs in order to identify all the logs related to an error, to a request
Here is a sub task from https://github.com/apollographql/router/issues/1840

The original idea was to provide a way to identify logs related to an error received in a GraphQL response with for example an `error_id`. I think we can do better by introduce a `request_id` in logs instead of just an `error_id`. In that way people who wants to debug a behavior without an error would be able to easily debug just by grepping the `request_id` in the logs.

My proposal is to introduce `request_id` in fields for JSON formatter and directly in the text with of the `TextFormatter` (including nice formatting). 

~~An addition to this experience we could introduce a way to directly ask to the router to return the `request_id` in the response thanks to GraphQL extensions to not limit it only for error cases. The way I imagine it is a configuration to indicate if we want to enable this debug mode and what kind of header should we pass to enable the ability to return the `request_id`.~~

An addition to this experience we could introduce a way to directly ask to the router to return the `request_id` in the response headers.

Something like

```yaml
expose_request_id:
    enabled: true # Disabled by default
    header_name: "ROUTER_DEBUG_MODE" # Send ROUTER_DEBUG_MODE = true in headers to return the request_id in extensions
```

and have an answer like this:

```json
{
  "data": {
    "topProducts": [
      {
        "name": "Table"
      },
      {
        "name": "Couch"
      },
      {
        "name": "Chair"
      }
    ]
  }
}
```

headers:

```
apollo_request_id: b060011c6c016070fc21a6c543dda554
```

And the formatted text log for example:

```log
2022-10-21T14:49:11.557950Z ERROR [request_id=b060011c6c016070fc21a6c543dda554] apollo_router::query_planner::execution: Fetch error: HTTP fetch failed from 'accounts': HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: Connection refused (os error 111)
```

TODO:

- [x] Rename to trace_id
- [x] Send it in apollo studio in response headers

### PR DESCRIPTION
**Pull request by bnjjj**
close #1981 

It automatically adds a `trace_id` on logs to identify which log is related to a specific request. Also adds `trace_id` to an example error to have an example. 

This PR also refactor the way we handled formatters before, this refactor will let us be more flexible on which specific fields we want to display in logs. For now it keeps the current behavior which do not display any fields coming from current span and parent spans, only fields written directly in the log macro.

Example of an error response:
```json
{
  "data": {},
  "errors": [
    {
      "message": "HTTP fetch failed from 'accounts': HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: Connection refused (os error 111)",
      "path": [],
      "extensions": {
        "type": "SubrequestHttpError",
        "service": "accounts",
        "reason": "HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: Connection refused (os error 111)"
      }
    }
  ]
}
```

with response header:
```
apollo_trace_id: 5e6a6bda8d0dca26e5aec14dafa6d96f
```

with related logs:

```logs
2022-10-21T15:17:45.562553Z ERROR [trace_id=5e6a6bda8d0dca26e5aec14dafa6d96f] apollo_router::services::subgraph_service: fetch_error="hyper::Error(Connect, ConnectError(\"tcp connect error\", Os { code: 111, kind: ConnectionRefused, message: \"Connection refused\" }))"
2022-10-21T15:17:45.565768Z ERROR [trace_id=5e6a6bda8d0dca26e5aec14dafa6d96f] apollo_router::query_planner::execution: Fetch error: HTTP fetch failed from 'accounts': HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: Connection refused (os error 111)
```

### CODE REVIEW

**Review by Geal — APPROVED** [review]

File: NEXT_CHANGELOG.md [hunk 1]
```diff
@@ -27,6 +27,20 @@ By [@USERNAME](https://github.com/USERNAME) in https://github.com/apollographql/
...
+
+### Add `trace_id` in logs to identify all logs related to a specific request [Issue #1981](https://github.com/apollographql/router/issues/1981))
+
+It automatically adds a `trace_id` on logs to identify which log is related to a specific request. Also adds `apollo_trace_id` in response headers to help the client to identify logs for this request.
+
+Example of logs in text:
+
+```logs
+2022-10-21T15:17:45.562553Z ERROR [trace_id=5e6a6bda8d0dca26e5aec14dafa6d96f] apollo_router::services::subgraph_service: fetch_error="hyper::Error(Connect, ConnectError(\"tcp connect error\", Os { code: 111, kind: ConnectionRefused, message: \"Connection refused\" }))"
+2022-10-21T15:17:45.565768Z ERROR [trace_id=5e6a6bda8d0dca26e5aec14dafa6d96f] apollo_router::query_planner::execution: Fetch error: HTTP fetch failed from 'accounts': HTTP fetch failed from 'accounts': error trying to connect: tcp connect error: Connection refused (os error 111)
```
Comment (garypen) [inline NEXT_CHANGELOG.md:39]: Could we have a JSON example as well?

File: apollo-router/src/plugins/telemetry/formatters/json.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+        let mut v = JsonVisitor::new(&mut writer);
+        fields.record(&mut v);
+        v.finish()
+    }
+
+    /// Record additional field(s) on an existing span.
+    ///
+    /// By default, this appends a space to the current set of fields if it is
+    /// non-empty, and then calls `self.format_fields`. If different behavior is
+    /// required, the default implementation of this method can be overridden.
```
Comment (garypen) [inline apollo-router/src/plugins/telemetry/formatters/json.rs:54]: I don't think this comment is correct. Isn't this the comment from the plain text formatter?

**Review by garypen — APPROVED** [review]
Nice work. It makes me a bit sad that we are internalising so much of the tracing formatting code, but hey ho...

Couple of small questions.

File: apollo-router/src/plugins/telemetry/formatters/json.rs (hunk 2 above)
Comment (bnjjj) [reply apollo-router/src/plugins/telemetry/formatters/json.rs:54]: Let me check, I copy pasted some parts of the tracing code

**Comment by bnjjj** [comment]
@garypen I agree with you about internalising some tracing formatting code. I'm still pretty happy because I didn't had to copy paste the biggest part of json formatter, only small parts. Also that's why [I started a draft PR in tracing](https://github.com/tokio-rs/tracing/pull/2353) to be able to implement filtering upstream and get rid of that part of the code.

File: apollo-router/src/plugins/telemetry/formatters/json.rs (hunk 2 above)
Comment (garypen) [reply apollo-router/src/plugins/telemetry/formatters/json.rs:54]: It looks like they copied that comment from the interface and it's wrong (I think) there as well. Never mind.

**Pull request merged**